Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,31 @@
+# create a lock using `flock (2)`. A name is required as first argument.
+# the lock will be automatically unlocked when the shell process quits.
+lock() {
+  local name="$1"
+  local lock_file="/tmp/homebrew-${name}.lock"
```

Comment:
My main issue with this, but that's unlikely to impact many users, is that there's no distinction between different Homebrew installations on the same machine. Once this is merged I won't be able to fire off a `brew update` in all of them simultaneously, or will I?